### PR TITLE
Implement testkit without `NodeHandler` usage.

### PR DESCRIPTION
**Pull request by alekseysidorov**

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -370,33 +520,101 @@ impl TestHarness {
...
+        if let Some(cfg_proposal) = self.cfg_proposal.take() {
+            match cfg_proposal {
+                Uncommitted(cfg_proposal) => {
+                    // Commit configuration proposal
+                    let stored = cfg_proposal.stored_configuration().clone();
+                    let mut fork = self.blockchain.fork();
+                    CoreSchema::new(&mut fork).commit_configuration(stored);
+                    let changes = fork.into_patch();
+                    self.blockchain.merge(changes).unwrap();
+                    self.cfg_proposal = Some(ConfigurationProposalState::Committed(cfg_proposal));
```
Comment (slowli) [inline src/lib.rs]: Nitpick: `ConfigurationProposalState::` qualifier could be skipped here, too.

File: src/lib.rs [hunk 2]
```diff
@@ -370,33 +520,101 @@ impl TestHarness {
...
+                }
+                Committed(ref cfg_proposal) if cfg_proposal.actual_from() == height => {
+                    // Modify the self configuration
+                    self.network_mut().update(
+                        cfg_proposal.us.clone(),
+                        cfg_proposal.validators.clone(),
+                    );
+                }
+                Committed(cfg_proposal) => {
+                    self.cfg_proposal = Some(ConfigurationProposalState::Committed(cfg_proposal));
```
Comment (slowli) [inline src/lib.rs]: ...and here.

File: src/lib.rs [hunk 3]
```diff
@@ -370,33 +520,101 @@ impl TestHarness {
...
+                }
+            }
+        }
+    }
+
+    /// Creates block with the given transactions.
+    /// Transactions that are in mempool will be ignored.
+    pub fn create_block_with_transactions(&mut self, txs: Vec<Box<Transaction>>) {
```
Comment (slowli) [inline src/lib.rs]: Repeating question from #2: what if some of transactions supplied to the method are already committed (i.e., present in one of previous blocks)?

File: src/lib.rs [hunk 4]
```diff
@@ -424,15 +640,189 @@ impl TestHarness {
...
+    pub fn network(&self) -> &TestNetwork {
+        &self.service_context.network
+    }
+
+    /// Returns the mutable reference to test network for manual modifications.
+    pub fn network_mut(&mut self) -> &mut TestNetwork {
+        &mut self.service_context.network
+    }
+
+    /// Returns the actual configuration of the testkit for modification.
```
Comment (slowli) [inline src/lib.rs]: Nitpick: the description could be a little misleading; it looks like this method would return a mutable ref to `TestNetworkConfiguration` actually used by the node. I would change the description to

> Returns a copy of the actual configuration of the testkit. The returned configuration could be modified for use with `propose_configuration_change()` method.

File: src/lib.rs [hunk 5]
```diff
@@ -424,15 +640,189 @@ impl TestHarness {
...
+        }
+    }
+
+    /// Returns the testkit node.
+    pub fn us(&self) -> &TestNode {
+        &self.us
+    }
+
+    /// Modifies the testkit node.
+    pub fn set_us(&mut self, us: TestNode) {
```
Comment (slowli) [inline src/lib.rs]: Maybe, allow to chain `set_*` methods by returning `&mut Self`? Although I suppose it wouldn't be completely in line with best practices, as `propose_configuration_change` consumes `TestNetworkConfiguration` instead of taking it by reference.

**Review by slowli — DISMISSED** [review]
Looks good for merging as of f8728af8f72928f28dc4c1f15efbd5d5bfb086c0.

File: src/lib.rs (hunk 3 above)
Comment (alekseysidorov) [reply src/lib.rs]: Hm, in my opinion that would raises panic.

File: src/lib.rs [hunk 6]
```diff
@@ -131,182 +185,213 @@ impl Validator {
...
+            us,
+        }
     }
 
-    /// Sets the validator count to be used in the harness emulation.
-    pub fn validators(&mut self, validator_count: u16) -> &mut Self {
-        assert!(
-            validator_count > 0,
-            "Number of validators should be positive"
+    /// Adds an additional validators.
```
Comment (slowli) [inline src/lib.rs]: Nitpick: "an" article is out of place here; I'd rephrase

> Sets the number of validator nodes in the test network.

File: src/lib.rs [hunk 7]
```diff
@@ -131,182 +185,213 @@ impl Validator {
...
+    pub fn with_validators(mut self, validators_count: u16) -> Self {
+        assert!(validators_count > 0, "At least one validator must be.");
```
Comment (slowli) [inline src/lib.rs]: Nitpick: consider changing the message to "At least one validator should be present in the network").

File: src/lib.rs [hunk 8]
```diff
@@ -370,33 +455,107 @@ impl TestHarness {
...
+            );
+
+            let mut tx_hashes = Vec::with_capacity(txs.len());
+            let snapshot = self.snapshot();
+            let schema = CoreSchema::new(&snapshot);
+            for tx in txs {
+                let txid = tx.hash();
+                assert!(
+                    !schema.transactions().contains(&txid),
+                    "Given transaction is already committed: {:?}",
```
Comment (slowli) [inline src/lib.rs]: Nitpick: "Given" seems redundant here.

File: src/lib.rs [hunk 9]
```diff
@@ -625,13 +980,23 @@ impl HarnessApi {
...
 }
+
+#[test]
+fn test_create_block_heights() {
+    let mut testkit = TestKitBuilder::validator().create();
+    assert_eq!(Height(1), testkit.current_height());
+    testkit.create_block();
+    assert_eq!(Height(2), testkit.current_height());
+    testkit.create_blocks_until(Height(6));
+    assert_eq!(Height(7), testkit.current_height());
```
Comment (slowli) [inline src/lib.rs]: Hm, I'd think that:
- `TestKit.current_height()` would return the height of the blockchain (i.e., the number of blocks in it). This seems to hold in the current implementation.
- `create_blocks_until(h)` should create blocks until the blockchain reaches height `h`; i.e., after the operation, the blockchain height would be exactly equal to `h`. This seems not to hold in the current implementation; instead, after the operation, the blockchain height is equal to `h + 1`.

That is, the following test would hold:

```rust
let h = testkit.current_height();
testkit.create_blocks_until(h);
assert_eq!(h, testkit.current_height());
```

**Review by slowli — COMMENTED** [review]
Couple of nitpicks as of 7602750e241c64ac99cb8ba547312249b89a0310.

File: src/lib.rs (hunk 7 above)
Comment (alekseysidorov) [reply src/lib.rs]: Fixed.

File: src/lib.rs (hunk 8 above)
Comment (alekseysidorov) [reply src/lib.rs]: Fixed

File: src/lib.rs (hunk 9 above)
Comment (alekseysidorov) [reply src/lib.rs]: I implemented `last_block_height` for height of latest committed block.

**Review by stanislav-tkach — DISMISSED** [review]

**Review by skletsun — DISMISSED** [review]

File: examples/timestamping.rs [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+use exonum::crypto::{gen_keypair, PublicKey};
+use exonum::blockchain::{Block, Schema, Service, Transaction};
+use exonum::messages::{FromRaw, Message, RawTransaction};
+use exonum::storage::Fork;
+use exonum::encoding;
+use exonum_testkit::{ApiKind, TestKitBuilder};
+
+// Simple service implementation.
+
+const SERVICE_ID: u16 = 1;
```
Comment (stanislav-tkach) [inline examples/timestamping.rs]: I propose to use different ids in the examples because users often create services with "reserved" ids.

**Review by stanislav-tkach — DISMISSED** [review]

**Review by stanislav-tkach — DISMISSED** [review]

File: src/lib.rs [hunk 11]
```diff
@@ -370,48 +606,123 @@ impl TestHarness {
...
 
-        let propose = self.network.us().create_propose(
-            height,
-            &last_hash,
-            tx_hashes,
-        );
-        let precommits: Vec<_> = self.network
+        let (block_hash, patch) = {
+            let validator_id = self.leader().validator_id().expect(
+                "Tested node is not a validator",
```
Comment (slowli) [inline src/lib.rs]: Nitpick: is this `expect()` still actual? It seems `TestKit.leader()` always returns (the first) validator, not the currently tested node.

File: src/lib.rs [hunk 12]
```diff
@@ -370,48 +606,123 @@ impl TestHarness {
...
             .iter()
             .map(|v| v.create_precommit(&propose, &block_hash))
             .collect();
-        handler.commit(block_hash, precommits.iter(), Some(round));
+
+        self.blockchain
+            .commit(&patch, block_hash, precommits.iter())
+            .unwrap();
+
+        self.poll_events();
```
Comment (slowli) [inline src/lib.rs]: What purpose does `poll_events()` serve here?

File: src/lib.rs [hunk 13]
```diff
@@ -370,48 +606,123 @@ impl TestHarness {
...
+    }
+
+    /// Update test network configuration if such an update has been scheduled
+    /// with `commit_configuration_change`.
+    fn update_configuration(&mut self) {
+        use ConfigurationProposalState::*;
+
+        let height = self.height().next().next();
```
Comment (slowli) [inline src/lib.rs]: Nitpick: I'd add a comment explaining why `height` needs to calculated as `self.height() + 2`; this is not obvious.

File: src/lib.rs [hunk 14]
```diff
@@ -370,48 +606,123 @@ impl TestHarness {
...
+                Uncommitted(cfg_proposal) => {
+                    // Commit configuration proposal
+                    let stored = cfg_proposal.stored_configuration().clone();
+                    let mut fork = self.blockchain.fork();
+                    CoreSchema::new(&mut fork).commit_configuration(stored);
+                    let changes = fork.into_patch();
+                    self.blockchain.merge(changes).unwrap();
+                    self.cfg_proposal = Some(Committed(cfg_proposal));
+                }
+                Committed(ref cfg_proposal) if cfg_proposal.actual_from() == height => {
```
Comment (slowli) [inline src/lib.rs]: Nitpick: I wonder whether it would make sense to take the condition into the body of the arm:

```rust
Committed(cfg_proposal) => {
    if cfg_proposal.actual_from() == height {
        // ...
    } else {
        // ...
    }
}
```

It could prevent cloning network params below. I'm not quite sure which way is more Rustic, of course 🤔

File: src/lib.rs [hunk 15]
```diff
@@ -424,15 +735,277 @@ impl TestHarness {
...
+    ///             .previous_configuration()
+    ///             .unwrap()
+    ///             .hash(),
+    ///         stored.previous_cfg_hash
+    ///     );
+    /// }
+    /// ```
+    pub fn commit_configuration_change(&mut self, proposal: TestNetworkConfiguration) {
+        use self::ConfigurationProposalState::*;
+        assert!(self.height() < proposal.actual_from());
```
Comment (slowli) [inline src/lib.rs]: Nitpick: could use descriptive messages for `assert!`s here and in some other places. May be done in a separate PR, though.

File: Cargo.toml [hunk 16]
```diff
@@ -1,10 +1,10 @@
 [package]
 name = "exonum-testkit"
-version = "0.0.0"
+version = "0.1.0"
```
Comment (slowli) [inline Cargo.toml:3]: Is there a particular reason to change the crate version? As far as I can see, it's not yet published, so I'd prefer to bump the crate version on dedicated tagged commits concurrently with publishing to `crates.io`. (At least, this seems to be the best practice with JS / Npm; maybe, Rust has other conventions?)

**Review by slowli — DISMISSED** [review]
Looks mergeable as of a5be7817b3fe624af675679aa8eca51fb229e21c. Due to the size of the PR, imo, there is a high probability of some unnoticed bugs or imperfections, but they could be fixed later.

Tiny nitpick about commit structure: could you please remove periods at the end of commit summaries (e.g., a9b7888398778e46c66563fc08bf0ce4a0b5e771)? Also, I'd try to squash "Format code" commits into others, but I understand it may get really tedious.

**Review by aleksuss — DISMISSED** [review]

File: src/lib.rs (hunk 12 above)
Comment (alekseysidorov) [reply src/lib.rs]: Because code below invokes `handle_commit` what can entail to the occurrence of events with transactions.
```rust
        self.blockchain
            .commit(&patch, block_hash, precommits.iter())
            .unwrap();
```

File: src/lib.rs (hunk 13 above)
Comment (alekseysidorov) [reply src/lib.rs]: Our `Height` wrapper lacks of `Add` trait implementation, @DarkEld3r can explain more detail.

File: src/lib.rs (hunk 15 above)
Comment (alekseysidorov) [reply src/lib.rs]: Fixed

File: src/lib.rs (hunk 14 above)
Comment (alekseysidorov) [reply src/lib.rs]: Fixed.

File: src/lib.rs (hunk 13 above)
Comment (slowli) [reply src/lib.rs]: I meant why it's `self.height() + 2` rather than `self.height() + 1` or something else; I have no complaints about the `.next().next()` construction.

**Review by slowli — APPROVED** [review]
Looks fine as of 78b22b42cab7f9451ef63056de2623281e9c0bf9.
